### PR TITLE
Dataloader optimisation, bug fix + ColourAugment

### PR DESCRIPTION
**Pull request by mikaelsimard5**
* Dataloader/Dataloader.py updated:
    - Sampling scheme reverted to what was functional.
    - Faster LoadFileParameter with list -> then pd.concat()
    - Corrected QueryImageFromCriteria.py to deal with queries resulting in no result.
    - Adjusted automatically num_workers to be 80% of what is available. 
* Inference/Image_Classifier.py
    - Made sure we don’t overwrite the pre-processed data when writing tumour tile probabilities.
    - Adjusted automatically num_workers to be 80% of what is available. 
* Dataloader/ObjectDetection.py updated.
* Inference/MitoticCell_Classifier.py added.
* Inference/ObjectDetection.py updated.
* Model/ConvNet.py
    - Confusion matrix added
* QA/Normalization/Colour/ColourAugment.py added
* QA/Normalization/Colour/ColourNorm.py has been optimised, with an option for colour jitter. Not as good as Colour Augment.
* Trainer/Image_Classifier.py
    - Adjusted trainer to use colour augmentation, not colour normalisation.
* Trainer/Image_Classifier_MIL.py: removed, it was not used and just a code snippet.
* Trainer/MitoticCell_Classifier.py added.

### CODE REVIEW

File: Dataloader/Dataloader.py [hunk 1]
```diff
@@ -35,78 +37,73 @@ def __getitem__(self, id):
...
-        gss                       = GroupShuffleSplit(n_splits=1, test_size=test_size, random_state=42)
-        train_val_inds, test_inds = next(gss.split(tile_dataset, groups=tile_dataset['id_external']))
-        tile_dataset_train_val    = tile_dataset.iloc[train_val_inds].reset_index()
-        tile_dataset_test         = tile_dataset.iloc[test_inds].reset_index()
+        tile_dataset['id_external'] = tile_dataset['id_external'].apply(str)
 
-        # Select only N per sample from each class
-        tile_dataset_train_val   = tile_dataset_train_val.groupby(target).head(n=n_per_sample)
-        tile_dataset_test        = tile_dataset_test.groupby(target).head(n=n_per_sample)
+        tile_dataset_sampled = sampling_schemes.sample_N_per_WSI(tile_dataset, n_per_sample=n_per_sample)
```
Comment (cacof1) [inline Dataloader/Dataloader.py:70]: What's the problem with the groupby?

File: Inference/Image_Classifier.py [hunk 2]
```diff
@@ -23,10 +23,15 @@
...
+print('Loading file parameters...', end='')
 tile_dataset = LoadFileParameter(config, SVS_dataset)
+tile_dataset_full = tile_dataset.copy()  # keep the full tile_dataset for final saving, but only process the reduced.
+valid_tumour_tiles_index = tile_dataset_full['prob_tissue_type_Tumour'] > 0.94
+tile_dataset = tile_dataset.loc[valid_tumour_tiles_index]
 
-# Mask the tile_dataset to only keep the tumour tiles, depending on a pre-set criteria.
-tile_dataset = tile_dataset[tile_dataset['prob_tissue_type_tumour'] > 0.85]
+# Assign SVS paths based on SVS_dataset (this will change depending on which workstation is running the script).
+tile_dataset['SVS_PATH'] = tile_dataset['id_external'].map(dict(zip(SVS_dataset.id_external, SVS_dataset.SVS_PATH)))
```
Comment (cacof1) [inline Inference/Image_Classifier.py:33]: Unclear what you are doing in here, isnt SVS_Path part of the Data Path in the Config?

File: Inference/ObjectDetection.py [hunk 3]
```diff
@@ -8,51 +8,158 @@
 os.environ["KMP_DUPLICATE_LIB_OK"]="TRUE"
 import cv2
 import toml
+from datetime import datetime, date
+import time
+import pathlib
+
 import torchvision.transforms as T
-from QA.Normalization.Colour import ColourNorm
+from QA.Normalization.Colour import ColourNorm_old
```
Comment (cacof1) [inline Inference/ObjectDetection.py:16]: Do we still have that?

File: QA/Normalization/Colour/ColourNorm.py [hunk 4]
```diff
@@ -3,101 +3,142 @@
...
 
-        # recreate the image using reference stain vectors
-        img[:, ids] = torch.exp(-torch.matmul(self.HERef, C))
-        img = img.reshape(c, h, w)
+        return OD, valid_idx
 
-        if self.get_stains:
-            H = torch.exp(torch.matmul(-self.HERef[:, 0].unsqueeze(-1), C[0, :].unsqueeze(0)))
-            H = H.T.reshape(c, h, w)
+    def compute_matrices(self, img, HE_test=None, maxC_test=None):
```
Comment (cacof1) [inline QA/Normalization/Colour/ColourNorm.py:88]: Can you rename this?

File: Training/NewScript.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# config = toml.load('/home/dgs/Dropbox/M/PostDoc/UCL/Code/Python/DigitalPathologyAI/Configs/preprocessing/infer_tumour_convnet_7classes.ini')
+config = toml.load('/Users/mikael/Dropbox/M/PostDoc/UCL/Code/Python/DigitalPathologyAI/Configs/sarcoma/trainer_sarcoma_convnet_15types.ini')
+
+########################################################################################################################
+# 1. Download all relevant files based on the configuration file
+
+SVS_dataset = QueryImageFromCriteria(config)
+SynchronizeSVS(config, SVS_dataset)
+DownloadNPY(config, SVS_dataset)
+print(SVS_dataset)
```
Comment (cacof1) [inline Training/NewScript.py:25]: Can you delete this file?

File: Training/ObjectDetection.py [hunk 6]
```diff
@@ -7,29 +7,34 @@
...
 
 config = toml.load(sys.argv[1])
-n_gpus = torch.cuda.device_count()
+n_gpus = len(config['MODEL']['GPU_ID'])
 
 augmentation = A.Compose([
     A.RandomCrop(width=config['DATA']['dim'][0], height=config['DATA']['dim'][1]),
     A.HorizontalFlip(p=config['AUGMENTATION']['horizontalflip']),
-    #A.RandomBrightnessContrast(p=config['AUGMENTATION']['randombrightnesscontrast']),
+    A.RandomBrightnessContrast(p=config['AUGMENTATION']['randombrightnesscontrast']),
```
Comment (cacof1) [inline Training/ObjectDetection.py:31]: Why use albumentations rather than torchvision?

File: Training/PreprocessingMIL_Zexi_Mar1.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
+from typing import Union
```
Comment (cacof1) [inline Training/PreprocessingMIL_Zexi_Mar1.py:1]: This file doesn't need to be committed until it's ready

Comment (cacof1) [reply Training/PreprocessingMIL_Zexi_Mar1.py:1]: That is to the main branch. Zexi can make a branch of her own

**Review by cacof1 — CHANGES_REQUESTED** [review]
Few changes that are mostly for flow and coherence

File: Dataloader/Dataloader.py (hunk 1 above)
Comment (mikaelsimard5) [reply Dataloader/Dataloader.py:70]: The sampling method created issues for the validation set for some reason, this needs to be investigated. Reverted back to the previous sampling method for now.

File: Training/ObjectDetection.py (hunk 6 above)
Comment (mikaelsimard5) [reply Training/ObjectDetection.py:31]: @SZY1234567 please advise

File: Training/NewScript.py (hunk 5 above)
Comment (mikaelsimard5) [reply Training/NewScript.py:25]: Done

File: Training/PreprocessingMIL_Zexi_Mar1.py (hunk 7 above)
Comment (mikaelsimard5) [reply Training/PreprocessingMIL_Zexi_Mar1.py:1]: Unstaged in new commit.

File: Inference/Image_Classifier.py (hunk 2 above)
Comment (mikaelsimard5) [reply Inference/Image_Classifier.py:33]: Moved to LoadFileParameter

File: Training/ObjectDetection.py (hunk 6 above)
Comment (SZY1234567) [reply Training/ObjectDetection.py:31]: albumentations can do transforms for the image and the paired mask

**Review by cacof1 — APPROVED** [review]
